### ISSUE TITLE
BUG: memory leak after creating arrays

### ISSUE THREAD

**Issue by JudahSchwartz** [numpy/numpy#22895]
### Describe the issue:

In a process the defines and then runs a method that creates many numpy arrays, some of the memory is not released.
With the following code example everything should be garbage collected by the time "stop the program?" is displayed, but even manually calling gc.collect leaves almost a gigabyte in memory that isnt cleaned up 

Result of `ps aux`
```
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         1 25.6  8.0 1232924 905788 pts/0  Ssl+ 14:36   1:21 /usr/bin/qemu-x86_64 /usr/local/bin/python /jfile/test.py

```

Results of docker stats

```
CONTAINER ID   NAME              CPU %     MEM USAGE / LIMIT    MEM %     NET I/O       BLOCK I/O     PIDS
058ae47c2ac9   exciting_edison   0.00%     876.5MiB / 10.7GiB   8.00%     1.02kB / 0B   0B / 16.4kB   6
```

### Reproduce the code example:

```python
def get_arr():
    arr = ['a', 'b', 'c']
    for i in range(23):
        arr += arr
    return arr


def my_method():
    import numpy as np

    arr = get_arr()
    [np.asarray(elt) for elt in arr]

my_method()

x = input("stop the program?")
```

### Runtime information:

numpy.__version__ > 1.24.1
sys.version > 3.10.9 (main, Dec  8 2022, 01:46:27) [GCC 10.2.1 20210110]
numpy.show_runtime() [{'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                      'found': ['SSSE3', 'SSE41', 'POPCNT', 'SSE42'],
                      'not_found': ['AVX',
                                    'F16C',
                                    'FMA3',
                                    'AVX2',
                                    'AVX512F',
                                    'AVX512CD',
                                    'AVX512_KNL',
                                    'AVX512_KNM',
                                    'AVX512_SKX',
                                    'AVX512_CLX',
                                    'AVX512_CNL',
                                    'AVX512_ICL']}},
 {'architecture': 'Core2',
  'filepath': '/usr/local/lib/python3.10/site-packages/numpy.libs/libopenblas64_p-r0-15028c96.3.21.so',
  'internal_api': 'openblas',
  'num_threads': 5,
  'prefix': 'libopenblas',
  'threading_layer': 'pthreads',
  'user_api': 'blas',
  'version': '0.3.21'}]
None

### Context for the issue:

This example is simply to reproduce the issue but this is causing our explanation to have very high unnecessary usage of memory

**Comment by rossbar** [comment]
How are you running this? For example, some interpreters like IPython might keep references to objects for namespace introspection or output caching.

**Comment by LeonaTaric** [comment]
> How are you running this? For example, some interpreters like IPython might keep references to objects for namespace introspection or output caching.

the array [np.asarray()] is not in namespace

**Comment by JudahSchwartz** [comment]
Im running in a docker image demisto/chromium:1.0.0.41207 which is based on python:3.10.9-slim-bullseye. You can see its definition here https://github.com/demisto/dockerfiles/blob/master/docker/chromium/Dockerfile
![image](https://user-images.githubusercontent.com/24831617/210164829-85dab51e-87c1-4c1a-8e9e-53fd38e77a1a.png)
This is CPython

**Comment by mattip** [comment]
Are you seeing a real memory leak or only a problem with growing RSS?

This is a known phenomenon with `malloc` and `free`: calling `free` does not necessarily return memory to the operation system, but it does mark it as available for another allocation inside the process. I think you are creating a worst-case scenario by allocating ever-larger pieces of memory. At some point the operating system will reclaim the freed pages. You can use an alternative memory allocation routine via [something like mimalloc](https://microsoft.github.io/mimalloc/index.html) or [use a NEP 49 memory management strategy](https://numpy.org/devdocs/reference/c-api/data_memory.html#configurable-memory-routines-in-numpy-nep-49) to work around this.

**Comment by seberg** [comment]
Yeah, you have to rerun `my_method()` a few times to proof a leak.  Not only malloc has memory arenas which might not be returned to the system, so does Python and you are creating a huge amount of small objects there.

This is a common surprise, but nothing suspicious.  Closing for now, thanks for reporting, please do comment if you disagree with the assessment (i.e. rerunning does not lead to reclaiming memory).

**Comment by JudahSchwartz** [comment]
will investigate that, thanks for the help!
